[Docs] Redshift Serverless Guide (#20168)
gravitational/teleport: docs/pages/database-access/guides/redshift-serverless.mdx

<ScopedBlock scope={["oss", "enterprise"]}>
![Teleport Database Access Redshift Self-Hosted](../../../img/database-access/guides/redshift_selfhosted_serverless.png)
</ScopedBlock>
<ScopedBlock scope={["cloud"]}>
![Teleport Database Access Redshift Cloud](../../../img/database-access/guides/redshift_cloud_serverless.png)
</ScopedBlock>

## Prerequisites

- A pre-existing Teleport `(=teleport.version=)` cluster. If you don't already
  have Teleport running, see [How to Choose a Teleport Edition](../../choose-an-edition/introduction.mdx)
  to learn more about your options, and how to set up Teleport.
- AWS account with a Redshift Serverless configuration and permissions to create
  and attach IAM policies.
- Command-line client `psql` installed and added to your system's `PATH` environment variable.
- A host where you will run the Teleport Database Service. This guide assumes an
  EC2 instance, and provides a corresponding example of access control.
- (!docs/pages/includes/tctl.mdx!)

## Step 1/5. Create an IAM policy for Teleport

<Admonition type="note" title="Example Access Control">
This guide provides an example configuration of IAM access roles as a model,

> and uses an EC2 instance to serve the Teleport Database Access service.

Reviewer: ```suggestion
and uses an EC2 instance to serve the Teleport Database Service.
```
To be consistent with the Core Concepts page (https://goteleport.com/docs/core-concepts/).

---
title: Database Access with Redshift Serverless on AWS
description: How to configure Teleport Database Access with AWS Redshift Serverless.
---

This guide will help you to:

> - Set up Teleport to access your AWS Redshift Serverless instances.

Reviewer 2: nitpick: not a big deal but since this is serverless it's a little bit weird to me that we say `instances` here. maybe `databases`, or `workgroups`?

Author: Sure, I was at a loss for terminology here.

> - Set up Teleport to access your AWS Redshift Serverless instances.

Author: ```suggestion
- Set up Teleport to access your AWS Redshift Serverless workgroups.
```

- Connect to your databases through Teleport.

<ScopedBlock scope={["oss", "enterprise"]}>
![Teleport Database Access Redshift Self-Hosted](../../../img/database-access/guides/redshift_selfhosted_serverless.png)
</ScopedBlock>
<ScopedBlock scope={["cloud"]}>
![Teleport Database Access Redshift Cloud](../../../img/database-access/guides/redshift_cloud_serverless.png)

> </ScopedBlock>

Reviewer 2: I think we should also include the STS service in these images (see the old dynamodb guide for example).

Author: I can do that. For my own edification, what is the STS service, and how does it relate to this guide?

Reviewer 2: sts is the Security Token Service that we make the AssumeRole call to impersonate the access role for the teleport user.

Author: Updated the images in a3d1684df72a0b5a16b3d08ef993f286847c1c37

## Prerequisites

> - A pre-existing Teleport `(=teleport.version=)` cluster. If you don't already

Reviewer: This Prerequisites section calls for a pre-existing cluster, but also (in `user-client-prereqs.mdx`) a host where you will install the Auth/Proxy.

Author: Replacing with `tctl.mdx`.

  have Teleport running, see [How to Choose a Teleport Edition](../../choose-an-edition/introduction.mdx)
  to learn more about your options, and how to set up Teleport.
- AWS account with a Redshift Serverless configuration and permissions to create
  and attach IAM policies.
- Command-line client `psql` installed and added to your system's `PATH` environment variable.
- A host where you will run the Teleport Database Service. This guide assumes an
  EC2 instance, and provides a corresponding example of access control.

> - (!docs/pages/includes/user-client-prereqs.mdx!)

Reviewer: For me, each bullet in this `Tabs` component is a continuation of the previous list (they're all black bullets at the same outline level), so we don't need to add this component as a bullet itself. With this extra bullet, the bullets within the `Tabs` component become indented/white.

## Step 1/5. Create an IAM policy for Teleport

<Admonition type="note" title="Example Access Control">

> This guide provides an example configuration of IAM access roles as a model,

Reviewer: ```suggestion
This guide provides an example configuration of IAM roles as a model,
```

> and uses an EC2 instance to serve the Teleport Database Access service.

Reviewer: ```suggestion
and uses an EC2 instance to serve the Teleport Database Service.
```

The level of access provided may not suit your needs, or may not fit your
organization's access conventions. You should adjust the AWS IAM permissions to
fit your needs.
</Admonition>

Teleport needs AWS IAM permissions to be able to:

> - Discover and register Redshift databases.

Reviewer 2: ```suggestion
- Discover and register Redshift Serverless databases.
```
or if we want to be more precise
```suggestion
- Discover and register Redshift Serverless workgroups and their associated VPC endpoints.
```

> - Manage IAM user or IAM role policies.

Reviewer: ```suggestion
- Manage IAM user and IAM role policies.
```

Author: Good catch. Gonna go even further:

```suggestion
- Manage IAM user and role policies.
```

If you don't already have one, create a role to apply to the EC2 instance that
the Teleport Database Service will run on. This guide uses the example name

> `teleport-redshift-node`.

Reviewer 2: nit
```suggestion
`teleport-redshift-serverless-node`.
```
I think it's better to have the serverless in the names to differentiate between the serverless vs the non-serverless. Same applies to other names.

Also, curious if we call these *node* or *server* or *host*?

Author: I wasn't sure myself to I reviewed some of our other docs and discussions, and it looks like *node* is still an appropriate term for a system running a Teleport access service but not the auth/proxy.

Did you also update the references to this renamed role, or should I go through and find/replace?

Reviewer 2: > I wasn't sure myself to I reviewed some of our other docs and discussions, and it looks like node is still an appropriate term for a system running a Teleport access service but not the auth/proxy.

Cool. I was just curious.

> Did you also update the references to this renamed role, or should I go through and find/replace?

No. You may need to find all the `redshift` without serverless and replace them.

Author: Updated throughout in a3d1684df72a0b5a16b3d08ef993f286847c1c37

Under **Trusted entity type** choose "AWS service", and under **Use case** choose "EC2".

Create and attach a policy to the `teleport-redshift-node` role to allow Teleport
to bootstrap access to AWS Redshift Serverless:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "iam:AttachRolePolicy",
                "iam:PutRolePermissionsBoundary"
            ],
            "Resource": [

>                 "arn:aws:iam::(=aws.aws_access_key=):role/teleport-redshift-node"

Reviewer 2: this should be an account ID?

Author: As mentioned in the discussion for the DynamoDB guide, the var `aws_access_key` is just a placeholder that returns a bunch of jibberish. It doesn't matter what the var name itself is, since the user won't see it, and nowhere in this guide do we need an example access key that would duplicate the content.

            ],
            "Effect": "Allow"
        },
        {
            "Action": [
                "iam:GetPolicy",
                "iam:CreatePolicy",
                "iam:TagPolicy",
                "iam:ListPolicyVersions",
                "iam:CreatePolicyVersion"
            ],

>             "Resource": [
>                 "*"
>             ],

Reviewer 2: If we want minimal permission:
```suggestion
            "Resource": [
                "arn:aws:iam::(account-id):policy/DatabaseAccess",
                "arn:aws:iam::(account-id):policy/DatabaseAccessBoundary"
            ],
```

Note that the poilcy name can be changed by `--policy-name` flag from the bootstrap command. Say `teleport db configure /opt/teleport/teleport_db.yaml --policy-name STeveDatabaseAccess`, then the policy name in the IAM policy would be
```
                "arn:aws:iam::(account-id):policy/STeveDatabaseAccess",
                "arn:aws:iam::(account-id):policy/STeveDatabaseAccessBoundary"
```

Author: Rephrasing to make sure I follow: You're saying:

- We can reduce the scope of these permissions by defining the policies this role can create. The policy names in your example match those created by our bootstrap command.
- In order for this to work, we must also update the bootstrap command we use later in the guide to specify this policy name.

Is that correct?

Reviewer 2: > We can reduce the scope of these permissions by defining the policies this role can create. The policy names in your example match those created by our bootstrap command.

Yes

> In order for this to work, we must also update the bootstrap command we use later in the guide to specify this policy name.

The `--policy-name` is an optional flag, if not specified, the value is `DatabaseAccess`. So we don't have to mention this, or maybe in a tip. Up to you what you think it's more appropriate.

Author: Updated and noted in a3d1684df72a0b5a16b3d08ef993f286847c1c37 - I elected not to discuss the `--policy-name` flag in this page.

            "Effect": "Allow"
        }
    ]
}
```

> Remember to replace the example AWS account ID.

Reviewer: This is the element in `Resource`?

<Admonition type="tip">

> This policy is required for Automatic Teleport bootstrapping, documented below.

Reviewer: Are all parts of this policy required for bootstrapping? If so, would it make sense to move this text to the top of the section so readers don't reach this tip only after creating and attaching the policy?

Another possibility would be to move these instructions into the "Automatic" tab in the "Bootstrap access" section, since (as far as I understand it) they only apply to that tab

Author: Going with your first suggestion, since it makes sense to do all the AWS things at once, rather than go back into the AWS console when we're talking about running CLI commands on the node.

You can optionally skip this policy, and select the "Manual" option when you
reach step 4.
</Admonition>

> Apply this role to the EC2 instance the Teleport Database Service will be

Reviewer: Is there a quick `aws` command we can use for this? If applying a role to an EC2 instance is something most readers will consider straightforward, though, maybe it's not something we need.

Author: I lack the AWS familiarity to answer this question. Pinging @greedy52 to see if they have an opinion.

> Apply this role to the EC2 instance the Teleport Database Service will be

Reviewer: ```suggestion
Apply the `teleport-redshift-node` role to the EC2 instance the Teleport Database Service will be
```
Since it's been a moment since we've mentioned the role by name

Author: Good call. Adjusting for the name change greedy52 asked for. Also doing it in a separate commit since it requires adjusting the line breaks.

installed in.

## Step 2/5. Create an IAM Role for user access

Create an AWS IAM role to provide user access to Redshift Serverless. This role
will be granted to Teleport users via a corresponding Teleport role. In this guide
we will use the example name `teleport-redshift-serverless-access`.

Under **Trusted entity type** choose "Custom trust policy". Edit the trust policy

> to allow the IAM role generated in the previous step to assume this role:

Reviewer: Maybe add a a short note about why the Database Service assumes this role?

                "Service": "ec2.amazonaws.com"
            },
            "Action": "sts:AssumeRole"
        }
    ]
}
```

Remember to replace the example AWS account ID.

Create and apply a permission policy to allow access to Redshift Serverless.
This policy can allow access to all instances:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "redshift-serverless:GetCredentials",
            "Resource": "*"
        }
    ]
}
```

Or you can restrict the `Resource` line to a specific Redshift Serverless workgroup:

```json
{
...
      "Resource": "arn:aws:redshift-serverless:us-west-2:1234567890:workgroup/some-workgroup-id"
...
}
```

## Step 3/5. Create a Teleport role for Redshift Serverless access

On your workstation logged in to your Teleport cluster with `tsh`, define a new
role to provide access to Redshift Serverless. Our example file is
`redshift-role.yaml`:

```yaml
version: v5
kind: role
metadata:
  name: redshift-serverless-access
spec:
  allow:
    db_labels:
      '*': '*'
    db_names:
    - dev
    db_users:

>     - 'teleport-redshift-serverless-access' # only the role name, do not specify the full AWS ARN

Reviewer 2: hmm.. this sounds more like a product bug. 

@smallinsky what do you think on this issue? Do we want to keep it "dumb" as it is (where `--db-user` must be one of `db_users`? Or should we allow mix-match between user name and ARNs? Or maybe force to specify ARN `--db-users`?

Reviewer 2: @smallinsky created #20359 to track this

Author: Sounds like we don't need to do any action or consider this a blocker, then?

Author: I'm going to assume "yes" to that question.

```

- The value of `db_users` corresponds to the IAM role created in the previous step.
- The value(s) for `db_names` will depend on your Redshift Serverless configuration,
  but `dev` is the default name applied by AWS. You can also provide `*` to grant
  access to all instances.

Save this file and apply it to your Teleport cluster:

```code
$ tctl create -f redshift-role.yaml
role 'redshift-serverless-access' has been created
```

> Add this role to your Teleport user. You can do this through the web UI, or by

Reviewer: Should we use the `add-role-to-user.mdx` partial here?

downloading your user definition:

```code
$ tctl get users/yourUser > yourUser.yaml
```

Adding the role:

```yaml
...
  roles:
  - redshift-serverless-access
...
```

And applying the changes:

```code
$ tctl create -f yourUser.yaml
user "yourUser" has been updated
```

## Step 4/5. Install and start the Teleport Database Service

(!docs/pages/includes/database-access/token.mdx!)

Install Teleport on the host where you will run the Teleport Database Service:

(!docs/pages/includes/install-linux.mdx!)

On the same host, create a Teleport configuration file:

```code
$ teleport db configure create \
   -o file \
   --proxy=teleport.example.com:443 \
   --token=/tmp/token \
   --redshift-serverless-discovery=us-west-1
```

> - Replace `teleport.example.com` with your Teleport proxy URI or cloud tenant

Reviewer: ```suggestion
- Replace `teleport.example.com` with your Teleport Proxy Service URI or cloud tenant
```

  address (e.g. `mytenant.teleport.sh`).
- Provide the token from the previous step to the `--token` flag.
- Update the value of `--redshift-serverless-discovery` to match your AWS region.

This command will generate a Database Service configuration with Redshift
Serverless auto-discovery enabled and place it at `/etc/teleport.yaml`.

<Admonition type="tip" title="Filter with AWS tags">

> Use `--aws-tags <key>=<value>,<key>=<value>` to only match those with specified

Reviewer: What does "those" refer to?

Author: I _think_ in this instance "those" would be "Redshift Serverless workgroups", but would like @greedy52 to confirm before I change it.

Reviewer 2: yes, "Redshift Serverless workgroups" in this context. Or "AWS resources" in general, as it applies to other discoveries.

Author: Updated 👍

AWS resource tags.

For example, if a workgroup has multiple VPC endpoints but
only one of the VPC endpoints will be accessed through Teleport, add the resource

> tag `teleport.dev/discovery` with value true to the VPC endpoint, and then specify

Reviewer: ```suggestion
tag `teleport.dev/discovery` with value `true` to the VPC endpoint, and then specify
```

Author: Thanks! Including this in another commit since I noticed that the line breaks in this section need formatting, and don't want merge conflicts.

`--aws-tags teleport.dev/discovery=true` when running `teleport db configure create`.
</Admonition>

### Bootstrap access to Redshift Serverless

Teleport can bootstrap IAM permissions for the Database Service based on its
configuration using the `teleport db configure bootstrap` command. You can use
this command in automatic or manual mode:

- In automatic mode, Teleport will attempt to create appropriate IAM policies
  and attach them to the specified IAM identity (user or role). This requires the
  IAM permissions we applied to the `teleport-redshift-node` role earlier.

> - In manual mode, Teleport will print required IAM policies. You can then create

Reviewer: ```suggestion
- In manual mode, Teleport will print the required IAM policies. You can then create
```

  and attach them manually using the AWS management console.

<Admonition type="note" title="AWS credentials">

> AWS Credentials are only required if you’re running the command in "automatic"

Reviewer: How do you supply AWS credentials? I only see that they are required. If the intention is to provide this information in the "Specifying Credentials" link, what information am I supposed to extract from the link?

Author: Clarifying in the next commit.

```code
$ teleport db configure bootstrap -c /etc/teleport.yaml --manual
```

</TabItem>
</Tabs>

See the full `bootstrap` command
[reference](../../database-access/reference/cli.mdx#teleport-db-configure-bootstrap)
for more information.

### Start the Database service

Run the following command on the Database Service node, depending on how you
installed Teleport:

<Tabs>
<TabItem label="Package Manager">

```code
$ sudo systemctl start teleport
```

</TabItem>
<TabItem label="TAR archive">

```code
$ sudo teleport install systemd --output=/etc/systemd/system/teleport.service;
$ sudo systemctl enable teleport
$ sudo systemctl start teleport
```

</TabItem>
</Tabs>

The Database Service will discover all Redshift databases according to the
configuration and register them in the cluster. The Database Service will also
attempt to configure IAM access policies for the discovered databases. Keep in
mind that AWS IAM changes may not propagate immediately and can take a few minutes
to come into effect.

<Admonition type="note" title="AWS credentials">

The Teleport Database Service uses the default credential provider chain to
find AWS credentials. See

> [Specifying Credentials](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials)

Reviewer: Is this a repeat of the "Specifying credentials" step from earlier?

Would it make more sense to have a step where you confirm that you have supplied the correct AWS credentials on the Database Service host? Also, since the host is an EC2 instance, should we tell the reader to use instance metadata for credentials?

Author: I think the former. This was brought in from another guide I used as a template, or written before the other admonition, I can't remember. In any case, I'm just gonna `rm` it.
